lib/auth/native/native.go

```diff
@@ -111,292 +105,74 @@ func precomputeKeys() {
-			roles, err := services.MarshalCertRoles(c.Roles)
-			if err != nil {
-				return nil, trace.Wrap(err)
-			}
-			cert.Permissions.Extensions[teleport.CertExtensionTeleportRoles] = roles
-		}
-		if c.RouteToCluster != "" {
-			cert.Permissions.Extensions[teleport.CertExtensionTeleportRouteToCluster] = c.RouteToCluster
-		}
-		if !c.ActiveRequests.IsEmpty() {
-			requests, err := c.ActiveRequests.Marshal()
-			if err != nil {
-				return nil, trace.Wrap(err)
-			}
-			cert.Permissions.Extensions[teleport.CertExtensionTeleportActiveRequests] = string(requests)
+		case privKey := <-keysChan:
+			privateKeys = append(privateKeys, privKey)
 		}
 	}
 
-	if err := cert.SignCert(rand.Reader, c.CASigner); err != nil {
-		return nil, trace.Wrap(err)
-	}
-	return ssh.MarshalAuthorizedKey(cert), nil
+	return privateKeys, nil
 }
 
-// BuildPrincipals takes a hostID, nodeName, clusterName, and role and builds a list of
-// principals to insert into a certificate. This function is backward compatible with
-// older clients which means:
-//   - If RoleAdmin is in the list of roles, only a single principal is returned: hostID
-//   - If nodename is empty, it is not included in the list of principals.
-func BuildPrincipals(hostID string, nodeName string, clusterName string, roles types.SystemRoles) []string {
-	// TODO(russjones): This should probably be clusterName, but we need to
-	// verify changing this won't break older clients.
-	if roles.Include(types.RoleAdmin) {
-		return []string{hostID}
-	}
-
-	// if no hostID was passed it, the user might be specifying an exact list of principals
-	if hostID == "" {
-		return []string{}
-	}
-
-	// always include the hostID, this is what teleport uses internally to find nodes
-	principals := []string{
-		fmt.Sprintf("%v.%v", hostID, clusterName),
-		hostID,
-	}
-
-	// nodeName is the DNS name, this is for OpenSSH interoperability
-	if nodeName != "" {
-		principals = append(principals, fmt.Sprintf("%s.%s", nodeName, clusterName))
-		principals = append(principals, nodeName)
-	}
-
-	// Add localhost and loopback addresses to allow connecting to proxy/host
-	// on the local machine. This should only matter for quickstart and local
-	// development.
-	principals = append(principals,
-		string(teleport.PrincipalLocalhost),
-		string(teleport.PrincipalLoopbackV4),
-		string(teleport.PrincipalLoopbackV6),
-	)
+// PrecomputeKeys sets this package into a mode where a small backlog of keys are
+// computed in advance. This should only be enabled if large spikes in key computation
+// are expected (e.g. in auth/proxy services). Safe to double-call.
+func PrecomputeKeys() {
+	startPrecomputeOnce.Do(func() {
+		go precomputeKeys()
+	})
+}
 
-	// deduplicate (in-case hostID and nodeName are the same) and return
-	return apiutils.Deduplicate(principals)
+// PrecomputeTestKeys generates RSA keys and reuse them to reduce CPU usage. This method should
+// only be in tests. Safe to call multiple times.
+func PrecomputeTestKeys() {
```

Comment:
nit: can we mark this function as Unsafe like:
```suggestion
func UnsafePrecomputeKeysInTests() {
```

Reply:
As we talked, I added `*testing.M` parameter to force the use of this function in tests.